javalanglib/src/main/scala/java/lang/System.scala

```diff
@@ -131,48 +123,129 @@ object System {
     } else (src match {
       case src: Array[AnyRef] =>
         dest match {
-          case dest: Array[AnyRef] => impl(src.length, dest.length, (i, j) => dest(i) = src(j))
-          case _                   => mismatch()
+          case dest: Array[AnyRef] =>
+            // bounds checks
+            src(srcPos)
+            src(srcPos + length)
+            dest(destPos)
+            dest(destPos + length)
```

Comment:
@sjrd if you know a better way to do this that will not reference the Scala runtime library, I'd be very happy :)

Reply:
I actually think that's a pretty clever way to do it :p

However there are a few issues I think:

* you forgot to check for `length < 0`, which you can do with:
  ```scala
  if (length < 0)
    src(length)
  ```
* you will incorrectly throw if `srcPos + length == src.length` or `destPos + length == dest.length`; you should use `src(srcPos + length - 1)` and `dest(destPos + length - 1)`
* when `length == 0`, you should also accept `srcPos == src.length` and `destPos == dest.length`

Overall, it seems easier to always do the test that was done before in `impl`, but if it's `true`, then do `src(-1)` to force throwing the appropriate exception. You can do that in `impl` without relying on the stdlib with:
```scala
if (srcPos < 0 || destPos < 0 || length < 0 ||
    srcPos > srcLen - length || destPos > destLen - length) {
  f(-1, -1) // throw an ArrayIndexOutOfBoundsException or UBE
}
```
This is not really efficient because it cannot be dead-code-eliminated when the exception is unchecked, but it doesn't matter, because this whole `arraycopy` implementation is by-passed when the optimizer is enabled (it's an intrinsic). We only need a *correct* implementation, not an efficient one.